kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/partitioner/AvroFieldValuePartitioner.java

```diff
@@ -73,26 +73,31 @@ private Optional<String> getPartitionKeyFromRecord(String topic, GenericRecord r
   private int calculatePartition(String topic, String key, int numPartitions) {
     PartitionGroupConfig groupConfig = this.getPartitionGroup(key);
     if (!this.parititonIteratorByTopicAndGroup.contains(topic, groupConfig)) {
-      List<Integer> availableTopicPartitions =
-          this.getAvailablePartitionsForTopic(topic, numPartitions);
-      int totalPartitions = availableTopicPartitions.size();
-      int fromIndex = (int) (groupConfig.getNormalizedFractionalStart() * totalPartitions);
-      int toIndex = (int) (groupConfig.getNormalizedFractionalEnd() * totalPartitions);
-      List<Integer> assignedPartitions = availableTopicPartitions.subList(fromIndex, toIndex);
-      LOG.info(
-          "topic: {}, group config: {}, member: {}, available partitions:{}, assigned partitions: {}",
-          topic,
-          groupConfig,
-          key,
-          availableTopicPartitions,
-          assignedPartitions);
-      // Using cyclic iterator
-      Iterator<Integer> partitionIterator = Iterables.cycle(assignedPartitions).iterator();
-      this.parititonIteratorByTopicAndGroup.put(topic, groupConfig, partitionIterator);
+      synchronized (parititonIteratorByTopicAndGroup) {
```

Comment:
Why is this required? I could see potential concurrency issues, but this doesn't seem like one of them. The only shared value is the table here, and if it's different topics/configs, we should support concurrent writes to different keys; if it's the same, we're generating the same iterator result anyway so an overwrite isn't a problematic.

In other words, this synchronization doesn't seem wrong but just unnecessary overhead.